plugins/woocommerce-admin/client/marketplace/components/install-flow/install-new-product-modal.tsx

```diff
@@ -0,0 +1,401 @@
+	const [ activateUrl, setActivateUrl ] = useState< string >();
+	const [ showModal, setShowModal ] = useState< boolean >( false );
+	const [ notice, setNotice ] = useState< {
+		message: string;
+		status: Status;
+	} >();
+
+	const query = useQuery();
+
+	// Check if the store is connected to Woo.com. This is run once, when the component is mounted.
+	useEffect( () => {
+		const wccomSettings = getAdminSetting( 'wccomHelper', {} );
+		const isStoreConnected = wccomSettings?.isConnected;
+
+		setInstalledProducts( wccomSettings?.installedProducts );
+
+		if ( isStoreConnected === false ) {
+			setInstallStatus( InstallFlowStatus.notConnected );
+			setNotice( {
+				status: 'warning',
+				message: __(
+					'In order to install a product, you need to first connect your account.',
+					'woocommerce'
+				),
+			} );
+		} else {
+			setInstallStatus( InstallFlowStatus.notInstalled );
+		}
+	}, [] );
+
+	/**
+	 * Listen for changes in the query, and show the modal if the installProduct query param is set.
+	 * If it's set, try to find the product in the products prop. We need it to be able to
+	 * display title, icon and send product ID to Woo.com to create an order.
+	 */
+	useEffect( () => {
+		setShowModal( false );
+		if ( ! query.installProduct ) {
+			return;
+		}
+
+		const productId = parseInt( query.installProduct, 10 );
+
+		/**
+		 * Try to find the product in the search results. We need to product to be able to
+		 * show the title and the icon.
+		 */
+		const productToInstall = props.products.find(
+			( item ) => item.id === productId
+		);
+
+		if ( ! productToInstall ) {
+			return;
+		}
+
+		if ( installedProducts ) {
+			const isInstalled = !! installedProducts.find(
+				( item ) => item === productToInstall.slug
+			);
+
+			if ( isInstalled ) {
+				return;
+			}
+		}
+
+		setShowModal( true );
+		setProduct( productToInstall );
+	}, [ query, props.products, installedProducts ] );
+
+	function activateClick() {
+		if ( ! activateUrl ) {
+			return;
+		}
+
+		setInstallStatus( InstallFlowStatus.activating );
+
+		recordEvent( 'marketplace_activate_new_product_clicked', {
+			product_id: product ? product.id : 0,
+		} );
+
+		fetch( activateUrl )
+			.then( () => {
+				setInstallStatus( InstallFlowStatus.activated );
+			} )
+			.catch( () => {
+				setInstallStatus( InstallFlowStatus.activationFailed );
+				setNotice( {
+					status: 'error',
+					message: __(
+						'Activation failed. Please try again from the plugins page.',
+						'woocommerce'
+					),
+				} );
+			} );
+	}
+
+	function orderAndInstall() {
+		if ( ! product || ! product.id ) {
+			return;
+		}
+
+		recordEvent( 'marketplace_install_new_product_clicked', {
+			product_id: product.id,
+		} );
+
+		setInstallStatus( InstallFlowStatus.installing );
+
+		createOrder( product.id )
+			.then( ( response ) => {
+				// This narrows the CreateOrderResponse type to CreateOrderSuccessResponse
+				if ( ! response.success ) {
+					throw response;
```

Comment:
Throwing this bubbles up the error and it caught by the catch block at the very end. That being said, I think code will never run, because when we're in the `then` block, `response.success` will always be true.

Reply:
Why not use `Promise. reject()`?